[Workflow] Added Function (and Twig extension) to retrieve a specific transition

epitre opened the pull request:
You can now easily retrieve a specific transition. Useful when you want the metadata of a specific transition.

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | hmmmm, should I create a ticket first ?
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/13907

I needed to get the metadata of a transition, but in order to do that, you need the transition object. So here is a PR to easily get the transition object

epitre commented:
Hi,
if someone can help me with the travis build failed.
I don't understand why it says that "Error: Call to undefined method Symfony\Component\Workflow\Workflow::getEnabledTransition()" as I did commit he method. You can see it here : https://github.com/symfony/symfony/pull/37428/files#diff-6d1e8dc0d2c0e7c4ab206d0efbb71516R241

and I don't understand all the other errors regarding src/Symfony/Component/Notifier/Bridge/FreeMobile, src/Symfony/Component/Notifier/Bridge/Slack, etc....
```PHP Fatal error:  Declaration of Symfony\Component\Notifier\Bridge\Slack\SlackTransport::doSend(Symfony\Component\Notifier\Message\MessageInterface $message): void must be compatible with Symfony\Component\Notifier\Transport\AbstractTransport::doSend(Symfony\Component\Notifier\Message\MessageInterface $message): Symfony\Component\Notifier\Message\SentMessage in /home/travis/build/symfony/symfony/src/Symfony/Component/Notifier/Bridge/Slack/SlackTransport.php on line 34```

I feel like it has nothing to do with this PR :/

noniagriconomie commented:
For the notifier code, recently the return signature of the method changed

For the workflow part, i do not know :s

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
+        $marking = $this->getMarking($subject);
+
+        foreach ($this->definition->getTransitions() as $transition) {
+            $transitionBlockerList = $this->buildTransitionBlockerListForTransition($subject, $marking, $transition);
```
You can do the transition name equal comparison here to avoid useless computation to retreive the good one

epitre replied:
Done !
Thank you :)

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
     }
 
+    /**
+     * {@inheritdoc}
```
It is not inside the interface thus no inherited doc here i fear

epitre replied:
you're right !
thank you :)

noniagriconomie commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
+    /**
+     * {@inheritdoc}
+     */
+    public function getEnabledTransition(object $subject, string $name)
```
As a new method you can add return typehint i think here

epitre replied:
yes !
Thank you :)

nicolas-grekas commented on src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:
```diff
@@ -64,6 +65,16 @@ public function getEnabledTransitions(object $subject, string $name = null): arr
+    /**
+     * Returns a transition or null.
+     *
+     * @return Transition|null Enabled transition asked for
```
the docblock should be removed entirely

epitre replied:
Done !

nicolas-grekas commented on src/Symfony/Bridge/Twig/Extension/WorkflowExtension.php:
```diff
@@ -64,6 +65,16 @@ public function getEnabledTransitions(object $subject, string $name = null): arr
+     */
+    public function getEnabledTransition(object $subject, string $transition, string $name = null): ?Transition
+    {
+        return $this->workflowRegistry->get($subject, $name)->getEnabledTransition($subject, $transition);
```
this means the bridge is now incompatible with workflow < 5.2, this should be reflected in the composer.json file, section "conflict" & "require-dev". That's the reason why deps=low tests fail.

epitre replied:
:pray:  Thank you.
I made the requested changes. Is it ok for you ?

I still have this error : ```1) Symfony\Component\Messenger\Bridge\Doctrine\Tests\Transport\DoctrineReceiverTest::testOccursRetryableExceptionFromConnection

Error: Class 'Doctrine\DBAL\Driver\PDOException' not found```

But I guess it has nothing to do with the Workflow component ^^

lyrixx commented on src/Symfony/Component/Workflow/Tests/WorkflowTest.php:
```diff
@@ -592,6 +592,32 @@ public function testGetEnabledTransitions()
+    {
+        $definition = $this->createComplexWorkflowDefinition();
+        $subject = new Subject();
+        $eventDispatcher = new EventDispatcher();
```
I don't think we need this complexity. Guard are already massively tested.
you can remove all things related to the Event Dispatcher + Guard?
It will make tests simpler to understand :) 
Thanks

epitre replied:
done !

lyrixx commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -235,6 +235,23 @@ public function getEnabledTransitions(object $subject)
+            }
+            $transitionBlockerList = $this->buildTransitionBlockerListForTransition($subject, $marking, $transition);
+            if ($transitionBlockerList->isEmpty()) {
+                return $transition;
```
What if many transition match?
cf https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L199

epitre replied:
:eyes: it's possible to have many transitions with the same name ? I did not realize...
In this case, I would have to add 2 optional arrays, one for the `froms` and one from the `tos`.
This way, it would be possible to identify specifically any transition.

I can try to do something, but I think it's overkill.

epitre replied:
So now, it is possible to retrieve any specific transition.
What do you think ?

lyrixx reviewed (commented):
Thanks for your contribution.

Instead of the current implementation, I would simplify a lot the code: 
You could leverage the current [`getEnabledTransitions()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L199) method in the bridge, without implementing a new method.

Less code == less bug :)

epitre commented:
> Thanks for your contribution.
> 
> Instead of the current implementation, I would simplify a lot the code:
> You could leverage the current [`getEnabledTransitions()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L199) method in the bridge, without implementing a new method.
> 
> Less code == less bug :)

@lyrixx That's actually what I did first (and that's what I currently do in my project). And then I thought it would make more sense to have a dedicated function...
What I need in a twig template, someone might need it in a service (or somewhere else ^^)

epitre commented:
Now, I have a problem with graphviz and PlantUML...

Maybe I should not have changed the complex workflow ^^

lyrixx commented on src/Symfony/Component/Workflow/Tests/WorkflowBuilderTrait.php:
```diff
@@ -19,6 +19,7 @@ private function createComplexWorkflowDefinition()
         $transitions[] = $transitionWithMetadataDumpStyle;
         $transitions[] = new Transition('t4', 'd', 'f');
         $transitions[] = new Transition('t5', 'e', 'g');
+        $transitions[] = new Transition('t3', 'f', 'e');
```
As you state, IMHO, it's a bad idea to change theses fixtures :) You don't need it
Could you revert theses changes?

epitre replied:
done !

lyrixx commented on src/Symfony/Component/Workflow/Workflow.php:
```diff
@@ -235,6 +235,33 @@ public function getEnabledTransitions(object $subject)
+    {
+        $marking = $this->getMarking($subject);
+
+        $result = array_filter(
```
I guess you were right, this is not really possible to have many transitions with the same name enabled at the very same time.  
I'm sorry for loosing your time, but could you revert this change to the one you had it yesterday? the previous version was more readable :)

Thanks

epitre replied:
done !

epitre committed:
Added Function (and Twig extension) to retrieve a specific transition

You can now easily retrieve a specific transition. Useful when you want the metadata of a specific transition.

lyrixx reviewed (approved):
Thanks for your first contributions :+1: 

I forgot to ask: Could you update both CHANGELOG and this will be perfect :yellow_heart:

epitre committed:
Update workflow Changelog

epitre commented:
@lyrixx i updated the workflow changelog.
What is the other changelog you want me to update ?

lyrixx commented:
The one in the bridge https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/CHANGELOG.md

epitre commented:
@lyrixx Done !

epitre committed:
Update Bridge Changelog

fabpot commented:
Thank you @epitre.
